### PR TITLE
Add update_cold_head

### PR DESCRIPTION
**Pull request by posvyatokum**
In infinite loop of cold storage update #7978 we will need to write head to BlockMisc. We shouldn't do it inside copying of one block, because technically blocks can be copied in arbitrary order. Also, we might want to copy blocks epoch by epoch and update head only after whole epoch. 
This PR is:
+ Adding `update_cold_head` function and integration test for it
+ Changing `cold_db` argument from `dyn Database` to `ColdDB` in cold storage functions
+ Making `NodeStorage`'s `ColdDB` generic for testing purposes. Thus adding Phantom field for the case when cold_store feature is not provided
+ Adding `NodeStorage::new_with_cold` for testing purposes
+ Adding `NodeStorage::cold_db` to access cold_storage as ColdDB

### CODE REVIEW

**Comment by posvyatokum** [comment]
@mm-near @jakmeier @Longarithm 
This is a PR about cold storage. All of you can take a look to warm up, but I assigned @jakmeier to review.

**Review by jakmeier — COMMENTED** [review]
Big thumbs-up for submitting small changes with tests as a separate PR! That makes it much easier for us to review.

I think the implementation looks good to me. `update_cold_head` is intentionally not used, yet, outside of tests, right?

I also left a comment regarding generic type parameter vs dynamic dispatch that I think we should make a clear decision now, as it will be annoying to change later.

File: core/store/src/cold_storage.rs [hunk 1]
```diff
@@ -84,30 +84,47 @@ fn copy_from_store(
...
-            }
+            transaction.set(col, key, value);
         }
     }
     cold_db.write(transaction)?;
     return Ok(());
 }
 
-pub fn test_cold_genesis_update(cold_db: &dyn Database, hot_store: &Store) -> io::Result<()> {
+pub fn update_cold_head<D: Database>(
```
[the remark this answers is not in the document]
Comment (posvyatokum) [reply core/store/src/cold_storage.rs:106]: Added comments. Comment about height finality went to the block copying function.

**Comment by posvyatokum** [comment]
> `update_cold_head` is intentionally not used, yet, outside of tests, right?

Yes. It soon will be used it testing tool, and only after that in some production code.

**Review by jakmeier — APPROVED** [review]

**Pull request merged**